From 7c88ff680226e38b71715d19789308f6c2ab8122 Mon Sep 17 00:00:00 2001
From: Jason Rogena <j.rogena@cgiar.org>
Date: Thu, 12 Jun 2014 14:33:22 +0300
Subject: [PATCH 194/213] Version 1.0.1: Editing farmer details

Almost done with editing farmer details. Need for a save point
since stuff might get real soon ;)
---
 .../org/cgiar/ilri/mistro/farmer/EditFarmer.java   | 127 +++++++++++++++++++--
 .../cgiar/ilri/mistro/farmer/FarmerSelection.java  |  70 +++++++++++-
 .../cgiar/ilri/mistro/farmer/LandingActivity.java  |   1 +
 .../org/cgiar/ilri/mistro/farmer/MainMenu.java     |  21 ++--
 .../ilri/mistro/farmer/backend/DataHandler.java    |   9 ++
 .../org/cgiar/ilri/mistro/farmer/carrier/Cow.java  |  23 ++--
 .../cgiar/ilri/mistro/farmer/carrier/Farmer.java   |   5 +-
 .../src/main/res/layout/activity_edit_farmer.xml   |  48 ++++----
 .../main/res/layout/activity_farmer_selection.xml  |   7 ++
 MistroFarmer/src/main/res/values/strings_en.xml    |   1 +
 MistroFarmer/src/main/res/values/strings_kp.xml    |   1 +
 MistroFarmer/src/main/res/values/strings_kr.xml    |   1 +
 MistroFarmer/src/main/res/values/strings_lu.xml    |   1 +
 MistroFarmer/src/main/res/values/strings_nn.xml    |   1 +
 MistroFarmer/src/main/res/values/strings_sw.xml    |   1 +
 15 files changed, 260 insertions(+), 57 deletions(-)

diff --git a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/EditFarmer.java b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/EditFarmer.java
index e5937f6..be50322 100644
--- a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/EditFarmer.java
+++ b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/EditFarmer.java
@@ -1,17 +1,21 @@
 package org.cgiar.ilri.mistro.farmer;
 
 import android.app.AlertDialog;
+import android.app.ProgressDialog;
 import android.content.Context;
 import android.content.DialogInterface;
 import android.content.Intent;
+import android.os.AsyncTask;
 import android.os.Bundle;
 import android.telephony.TelephonyManager;
+import android.util.Log;
 import android.view.View;
 import android.widget.ArrayAdapter;
 import android.widget.Button;
 import android.widget.EditText;
 import android.widget.Spinner;
 import android.widget.TextView;
+import android.widget.Toast;
 
 import com.actionbarsherlock.app.SherlockActivity;
 import com.actionbarsherlock.view.Menu;
@@ -20,14 +24,19 @@ import com.actionbarsherlock.view.MenuItem;
 
 import org.cgiar.ilri.mistro.farmer.backend.DataHandler;
 import org.cgiar.ilri.mistro.farmer.backend.Locale;
+import org.cgiar.ilri.mistro.farmer.backend.database.DatabaseHelper;
+import org.cgiar.ilri.mistro.farmer.carrier.Farmer;
+import org.json.JSONArray;
 
+import java.util.ArrayList;
 import java.util.List;
 
 
 public class EditFarmer extends SherlockActivity implements MistroActivity, View.OnClickListener {
+    private static final String TAG = "EditFarmer";
 
-    private Menu menu;
 
+    private Menu menu;
     private TextView fullNameTV;
     private EditText fullNameET;
     private TextView preferredLanguageTV;
@@ -36,12 +45,14 @@ public class EditFarmer extends SherlockActivity implements MistroActivity, View
     private Spinner extensionPersonnelS;
     private TextView mobileNumberTV;
     private EditText mobileNumberET;
-    private TextView numberOfCowsTV;
-    private EditText numberOfCowsET;
     private Button editB;
+    private Button cancelB;
 
+    private String adminData;
+    private Farmer farmer;
     private List<String> languages;
     private boolean cacheData;
+    private List<String> vetNames;
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
@@ -62,16 +73,15 @@ public class EditFarmer extends SherlockActivity implements MistroActivity, View
         TelephonyManager telephonyManager=(TelephonyManager)this.getSystemService(Context.TELEPHONY_SERVICE);
         //Toast.makeText(this,telephonyManager.getSimSerialNumber(),Toast.LENGTH_LONG).show();
         mobileNumberET.setText(telephonyManager.getLine1Number());
-        numberOfCowsTV=(TextView)this.findViewById(R.id.number_of_cows_tv);
-        numberOfCowsET=(EditText)this.findViewById(R.id.number_of_cows_et);
         editB=(Button)this.findViewById(R.id.edit_b);
         editB.setOnClickListener(this);
+        cancelB =(Button)this.findViewById(R.id.cancel_b);
+        cancelB.setOnClickListener(this);
 
         //init text according to locale
         initTextInViews();
     }
 
-
     public boolean onCreateOptionsMenu(com.actionbarsherlock.view.Menu menu) {
         MenuInflater inflater = getSupportMenuInflater();
         inflater.inflate(R.menu.edit_farmer, menu);
@@ -115,8 +125,14 @@ public class EditFarmer extends SherlockActivity implements MistroActivity, View
         setTitle(Locale.getStringInLocale("edit_farmer", this));
         fullNameTV.setText(" * " + Locale.getStringInLocale("full_name", this));
         extensionPersonnelTV.setText(Locale.getStringInLocale("extension_p", this));
-        mobileNumberTV.setText(" * " + Locale.getStringInLocale("mobile_number", this));
-        numberOfCowsTV.setText(" * "+Locale.getStringInLocale("number_of_cows",this));
+
+        List<String> tmpVetNames = new ArrayList<String>();
+        tmpVetNames.add("");
+        ArrayAdapter<String> epArrayAdapter = new ArrayAdapter<String>(this,android.R.layout.simple_spinner_item,tmpVetNames);
+        epArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
+        extensionPersonnelS.setAdapter(epArrayAdapter);
+
+        mobileNumberTV.setText(" * " + Locale.getStringInLocale("phone_number", this));
         preferredLanguageTV.setText(" * "+Locale.getStringInLocale("preferred_language", this));
 
         ArrayAdapter<String> languageAdapter = new ArrayAdapter<String>(this,android.R.layout.simple_spinner_item, languages);
@@ -124,6 +140,7 @@ public class EditFarmer extends SherlockActivity implements MistroActivity, View
         preferredLanguageS.setAdapter(languageAdapter);
 
         editB.setText(Locale.getStringInLocale("edit", this));
+        cancelB.setText(Locale.getStringInLocale("cancel", this));
     }
 
     private void initMenuText(){
@@ -133,6 +150,38 @@ public class EditFarmer extends SherlockActivity implements MistroActivity, View
         }
     }
 
+    @Override
+    protected void onResume() {
+        super.onResume();
+        Bundle bundle = this.getIntent().getExtras();
+        if(bundle != null){
+            adminData = bundle.getString(FarmerSelection.KEY_ADMIN_DATA);
+            farmer = bundle.getParcelable(Farmer.PARCELABLE_KEY);
+            if(farmer != null){
+                fullNameET.setText(farmer.getFullName());
+                mobileNumberET.setText(farmer.getMobileNumber());
+
+                String prefLanguage = Locale.getLanguage(this, farmer.getPreferredLocale());
+
+                for(int i = 0; i < languages.size(); i++){
+                    if(languages.get(i).equals(prefLanguage)){
+                        preferredLanguageS.setSelection(i);
+                    }
+                }
+
+                //fetch extension personnel names from server. Do this after farmer object is initialized since the thread is going to use this object
+                FetchVetsThread fetchVetsThread = new FetchVetsThread();
+                fetchVetsThread.execute(0);
+            }
+            else{
+                Log.e(TAG, "Parcelable farmer object from previous activity is null");
+            }
+        }
+        else{
+            Log.e(TAG, "Unable to get data from previous activity");
+        }
+    }
+
     private void cacheEditTextData(){
         if(cacheData) {
             DataHandler.setSharedPreference(this, DataHandler.SP_KEY_FRA_FULL_NAME, fullNameET.getText().toString());
@@ -155,6 +204,68 @@ public class EditFarmer extends SherlockActivity implements MistroActivity, View
 
     @Override
     public void onClick(View v) {
+        if(v.equals(cancelB)){
+            Intent intent = new Intent(this, FarmerSelection.class);
+            intent.putExtra(FarmerSelection.KEY_ADMIN_DATA, adminData);
+            startActivity(intent);
+        }
+    }
+
+    private class FetchVetsThread extends AsyncTask<Integer, Integer, String> {
+        private ProgressDialog progressDialog;
+
+        @Override
+        protected void onPreExecute() {
+            super.onPreExecute();
+            progressDialog = ProgressDialog.show(EditFarmer.this, "", Locale.getStringInLocale("loading_please_wait", EditFarmer.this), true);
+        }
+
+        @Override
+        protected String doInBackground(Integer... params) {
+            Log.d(TAG, "Fetching extension personnel from server");
+
+            return DataHandler.sendDataToServer(EditFarmer.this, "", DataHandler.FARMER_FETCH_VETS_URL, true, DatabaseHelper.TABLE_EXTENSION_PERSONNEL);
+        }
+
+        @Override
+        protected void onPostExecute(String result) {
+            super.onPostExecute(result);
+            progressDialog.dismiss();
 
+            if(result == null){
+                Toast.makeText(EditFarmer.this, Locale.getStringInLocale("unable_to_get_epersonnel", EditFarmer.this), Toast.LENGTH_LONG).show();
+            }
+            else{
+                try{
+                    Log.d(TAG, "result is "+result);
+                    JSONArray vetJsonArray = new JSONArray(result);
+                    vetNames = new ArrayList<String>();
+                    vetNames.add("");
+                    for(int i = 0; i < vetJsonArray.length(); i++){
+                        vetNames.add(vetJsonArray.getJSONObject(i).getString("name"));
+                    }
+
+                    int selection = 0;//default selection is the blank one
+                    if(farmer != null){
+                        String selectedEP = farmer.getExtensionPersonnel();
+                        if(selectedEP != null && selectedEP.length() > 0){
+                            for(int i =0; i < vetNames.size(); i++){
+                                if(vetNames.get(i).equals(selectedEP)){
+                                    selection = i;
+                                }
+                            }
+                        }
+                    }
+
+                    ArrayAdapter<String> epArrayAdapter = new ArrayAdapter<String>(EditFarmer.this, android.R.layout.simple_spinner_item, vetNames);
+                    epArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
+                    extensionPersonnelS.setAdapter(epArrayAdapter);
+                    extensionPersonnelS.setSelection(selection);
+                }
+                catch (Exception e){
+                    e.printStackTrace();
+                }
+            }
+        }
     }
 }
diff --git a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/FarmerSelection.java b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/FarmerSelection.java
index 78d250f..1584ac6 100644
--- a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/FarmerSelection.java
+++ b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/FarmerSelection.java
@@ -2,7 +2,9 @@ package org.cgiar.ilri.mistro.farmer;
 
 import android.content.Intent;
 import android.os.Bundle;
+import android.util.Log;
 import android.view.View;
+import android.widget.ArrayAdapter;
 import android.widget.Button;
 import android.widget.Spinner;
 import android.widget.TextView;
@@ -13,15 +15,28 @@ import com.actionbarsherlock.view.MenuInflater;
 import com.actionbarsherlock.view.MenuItem;
 
 import org.cgiar.ilri.mistro.farmer.backend.Locale;
+import org.cgiar.ilri.mistro.farmer.carrier.Farmer;
+import org.json.JSONArray;
+import org.json.JSONObject;
+
+import java.util.ArrayList;
+import java.util.List;
 
 
 public class FarmerSelection extends SherlockActivity implements MistroActivity, View.OnClickListener {
 
+    private static final String TAG = "FarmerSelection";
+    public static final String KEY_ADMIN_DATA= "adminData";
+
     private Menu menu;
 
     private TextView selectFarmerTV;
     private Spinner selectFarmerS;
     private Button selectB;
+    private Button backB;
+
+    private List<Farmer> farmers;
+    private JSONObject adminData;
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
@@ -33,10 +48,48 @@ public class FarmerSelection extends SherlockActivity implements MistroActivity,
 
         selectB = (Button)findViewById(R.id.select_b);
         selectB.setOnClickListener(this);
+        backB = (Button)findViewById(R.id.back_b);
+        backB.setOnClickListener(this);
 
         initTextInViews();
     }
 
+    @Override
+    protected void onResume() {
+        super.onResume();
+
+        Bundle bundle=this.getIntent().getExtras();
+        if(bundle != null){
+            String adminJSONString = bundle.getString(KEY_ADMIN_DATA);
+            try{
+                adminData = new JSONObject(adminJSONString);
+                JSONArray farmerData = adminData.getJSONArray("farmers");
+
+                farmers = new ArrayList<Farmer>(farmerData.length());
+                for(int i = 0; i < farmerData.length(); i++){
+                    String ePersonnel = "";
+                    if(!farmerData.getJSONObject(i).getString("extension_personnel_id").equals("NULL")){
+                        ePersonnel = adminData.getString("name");
+                    }
+                    Farmer currFarmer = new Farmer(farmerData.getJSONObject(i), ePersonnel);
+                    farmers.add(currFarmer);
+                }
+
+                List<String> farmerNames = new ArrayList<String>(farmers.size());
+                for(int i = 0; i < farmers.size(); i++){
+                    Farmer currFarmer = farmers.get(i);
+                    farmerNames.add(currFarmer.getFullName());
+                }
+
+                ArrayAdapter<String> farmerArrayAdapter = new ArrayAdapter<String>(this,android.R.layout.simple_spinner_item, farmerNames);
+                farmerArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
+                selectFarmerS.setAdapter(farmerArrayAdapter);
+            }
+            catch (Exception e){
+
+            }
+        }
+    }
 
     public boolean onCreateOptionsMenu(com.actionbarsherlock.view.Menu menu) {
         MenuInflater inflater = getSupportMenuInflater();
@@ -66,6 +119,7 @@ public class FarmerSelection extends SherlockActivity implements MistroActivity,
 
         selectFarmerTV.setText(Locale.getStringInLocale("select_farmer", this));
         selectB.setText(Locale.getStringInLocale("select", this));
+        backB.setText(Locale.getStringInLocale("back", this));
     }
 
     private void initMenuText(){
@@ -78,7 +132,21 @@ public class FarmerSelection extends SherlockActivity implements MistroActivity,
     @Override
     public void onClick(View v) {
         if(v.equals(selectB)){
-            Intent intent = new Intent(this, EditFarmer.class);
+            if(selectFarmerS.getSelectedItemPosition() != -1 && farmers.size() > selectFarmerS.getSelectedItemPosition()){
+                Log.d(TAG, "Selected farmer index = "+String.valueOf(selectFarmerS.getSelectedItemPosition()));
+                Farmer selectedFarmer = farmers.get(selectFarmerS.getSelectedItemPosition());
+                Intent intent = new Intent(this, EditFarmer.class);
+                Bundle bundle = new Bundle();
+                bundle.putParcelable(Farmer.PARCELABLE_KEY, selectedFarmer);
+                intent.putExtras(bundle);
+                intent.putExtra(KEY_ADMIN_DATA, adminData.toString());
+                startActivity(intent);
+            }
+        }
+        else if(v.equals(backB)){
+            Intent intent = new Intent(this, MainMenu.class);
+            intent.putExtra(MainMenu.KEY_MODE, MainMenu.MODE_ADMIN);
+            intent.putExtra(MainMenu.KEY_ADMIN_DATA, adminData.toString());
             startActivity(intent);
         }
     }
diff --git a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/LandingActivity.java b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/LandingActivity.java
index 4a942c8..3a53ea1 100644
--- a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/LandingActivity.java
+++ b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/LandingActivity.java
@@ -353,6 +353,7 @@ public class LandingActivity extends SherlockActivity implements MistroActivity,
                 Toast.makeText(LandingActivity.this, Locale.getStringInLocale("sim_card_not_admin", LandingActivity.this), Toast.LENGTH_LONG).show();
             }
             else{
+                Log.d(TAG, "Admin data from server = "+result);
                 Intent intent=new Intent(LandingActivity.this,MainMenu.class);
                 intent.putExtra(MainMenu.KEY_MODE, MainMenu.MODE_ADMIN);
                 intent.putExtra(MainMenu.KEY_ADMIN_DATA, result);
diff --git a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/MainMenu.java b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/MainMenu.java
index 76d1eb9..9435d2f 100644
--- a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/MainMenu.java
+++ b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/MainMenu.java
@@ -56,6 +56,7 @@ public class MainMenu extends SherlockActivity implements MistroActivity, View.O
     private LocationManager locationManager;
     private String longitude;
     private String latitude;
+    private List<Farmer> farmers;
 
 
     @Override
@@ -98,18 +99,10 @@ public class MainMenu extends SherlockActivity implements MistroActivity, View.O
                 }
                 else if(mode.equals(MODE_ADMIN)){
                     String adminJSONString = bundle.getString(KEY_ADMIN_DATA);
-
+                    Log.d(TAG, " *** Admin data from login screen = "+adminJSONString);
                     try{
                         adminData = new JSONObject(adminJSONString);
 
-                        JSONArray farmerJsonArray = adminData.getJSONArray("farmers");
-
-                        List<Farmer> farmers = new ArrayList<Farmer>(farmerJsonArray.length());
-                        for(int i = 0; i < farmerJsonArray.length(); i++){
-                            Farmer currFarmer = new Farmer(farmerJsonArray.getJSONObject(i), adminData.getString("name"));
-                            farmers.add(currFarmer);
-                        }
-
                         Toast.makeText(this, Locale.getStringInLocale("welcome", this) + " " + adminData.getString("name") + " (" + Locale.getStringInLocale("admin", this) + ")", Toast.LENGTH_LONG).show();
                     }
                     catch (Exception e){
@@ -189,8 +182,14 @@ public class MainMenu extends SherlockActivity implements MistroActivity, View.O
             startActivity(intent);
         }
         else if(view==editFarmerB){
-            Intent intent = new Intent(this, FarmerSelection.class);
-            startActivity(intent);
+            if(adminData != null){
+                Intent intent = new Intent(this, FarmerSelection.class);
+                intent.putExtra(FarmerSelection.KEY_ADMIN_DATA, adminData.toString());
+                startActivity(intent);
+            }
+            else {
+                Log.w(TAG, "Admin data is null. Unable to move to the next activity");
+            }
         }
         else if(view==logoutB){
             Intent intent = new Intent(this, LandingActivity.class);
diff --git a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/backend/DataHandler.java b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/backend/DataHandler.java
index cee49ce..ac5b85f 100644
--- a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/backend/DataHandler.java
+++ b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/backend/DataHandler.java
@@ -1071,4 +1071,13 @@ public class DataHandler
             return valid;
         }
     }
+
+    public static boolean isNull(String string){
+        if(string == null || string.equals("NULL") || string.equals("null")){
+            return true;
+        }
+        else{
+            return false;
+        }
+    }
 }
diff --git a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/carrier/Cow.java b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/carrier/Cow.java
index 7f244bf..9ded465 100644
--- a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/carrier/Cow.java
+++ b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/carrier/Cow.java
@@ -5,6 +5,7 @@ import android.os.Parcel;
 import android.os.Parcelable;
 import android.util.Log;
 
+import org.cgiar.ilri.mistro.farmer.backend.DataHandler;
 import org.cgiar.ilri.mistro.farmer.backend.Locale;
 import org.json.JSONArray;
 import org.json.JSONException;
@@ -107,8 +108,8 @@ public class Cow implements Parcelable, Serializable {
             JSONObject thisCow = allCows.getJSONObject(index);
             initFromJSON(thisCow);
 
-            int sireID = (thisCow.getString("sire_id").equals("NULL")) ? -1 : thisCow.getInt("sire_id");
-            int damID = (thisCow.getString("dam_id").equals("NULL")) ? -1 : thisCow.getInt("dam_id");
+            int sireID = (DataHandler.isNull(thisCow.getString("sire_id"))) ? -1 : thisCow.getInt("sire_id");
+            int damID = (DataHandler.isNull(thisCow.getString("dam_id"))) ? -1 : thisCow.getInt("dam_id");
 
             if(sireID != -1){
                 for(int i = 0; i < allCows.length(); i++){
@@ -117,14 +118,14 @@ public class Cow implements Parcelable, Serializable {
                         Sire sire = new Sire();
                         sire.setName(possibleSire.getString("name"));
                         sire.setEarTagNumber(possibleSire.getString("ear_tag_number"));
-                        sire.setOwner((possibleSire.getString("owner_name").equals("NULL")) ? "" : possibleSire.getString("owner_name"));
-                        sire.setOwnerType((possibleSire.getString("bull_owner").equals("NULL")) ? "" : possibleSire.getString("owner_name"));
+                        sire.setOwner((DataHandler.isNull(possibleSire.getString("owner_name"))) ? "" : possibleSire.getString("owner_name"));
+                        sire.setOwnerType((DataHandler.isNull(possibleSire.getString("bull_owner"))) ? "" : possibleSire.getString("owner_name"));
 
                         this.sire = sire;
                     }
                 }
             }
-            this.sire.setStrawNumber((thisCow.getString("straw").equals("NULL")) ? "" : thisCow.getString("straw"));
+            this.sire.setStrawNumber((DataHandler.isNull(thisCow.getString("straw"))) ? "" : thisCow.getString("straw"));
 
             if(damID != -1){
                 for(int i = 0; i < allCows.length(); i++){
@@ -138,7 +139,7 @@ public class Cow implements Parcelable, Serializable {
                     }
                 }
             }
-            this.dam.setEmbryoNumber((thisCow.getString("embryo").equals("NULL")) ? "" : thisCow.getString("embryo"));
+            this.dam.setEmbryoNumber((DataHandler.isNull(thisCow.getString("embryo"))) ? "" : thisCow.getString("embryo"));
         }
         catch (Exception e){
             e.printStackTrace();
@@ -149,10 +150,10 @@ public class Cow implements Parcelable, Serializable {
         try {
             name = jsonObject.getString("name");
             earTagNumber = jsonObject.getString("ear_tag_number");
-            dateOfBirth = (jsonObject.getString("date_of_birth").equals("NULL")) ? "" : jsonObject.getString("date_of_birth");
+            dateOfBirth = (DataHandler.isNull(jsonObject.getString("date_of_birth"))) ? "" : jsonObject.getString("date_of_birth");
             dateAdded = jsonObject.getString("date_added");
-            age = (jsonObject.getString("age").equals("NULL")) ? -1 : jsonObject.getInt("age");
-            ageType = (jsonObject.getString("age_type").equals("NULL")) ? "" : jsonObject.getString("age_type");
+            age = (DataHandler.isNull(jsonObject.getString("age"))) ? -1 : jsonObject.getInt("age");
+            ageType = (DataHandler.isNull(jsonObject.getString("age_type"))) ? "" : jsonObject.getString("age_type");
 
             JSONArray breedArray = jsonObject.getJSONArray("breed");
             this.breeds = new ArrayList<String>(breedArray.length());
@@ -174,7 +175,7 @@ public class Cow implements Parcelable, Serializable {
 
             mode = "";
             countryOfOrigin = "";//only set this if is sire
-            serviceType = (jsonObject.getString("service_type").equals("NULL")) ? "" : jsonObject.getString("service_type");
+            serviceType = (DataHandler.isNull(jsonObject.getString("service_type"))) ? "" : jsonObject.getString("service_type");
             otherBreed = "";
             piggyBack = "";
             this.events = new ArrayList<Event>();
@@ -184,7 +185,7 @@ public class Cow implements Parcelable, Serializable {
                 inCalf = false;
             }
             else if(this.sex.equals(SEX_FEMALE)){
-                milkingStatus = (jsonObject.getString("milking_status").equals("NULL")) ? MILKING_S_ADULT_MILKING : jsonObject.getString("milking_status");
+                milkingStatus = (DataHandler.isNull(jsonObject.getString("milking_status"))) ? MILKING_S_ADULT_MILKING : jsonObject.getString("milking_status");
                 if(jsonObject.getString("in_calf").equals("1")){
                     inCalf = true;
                 }
diff --git a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/carrier/Farmer.java b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/carrier/Farmer.java
index e3b4417..828d2ad 100644
--- a/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/carrier/Farmer.java
+++ b/MistroFarmer/src/main/java/org/cgiar/ilri/mistro/farmer/carrier/Farmer.java
@@ -54,11 +54,12 @@ public class Farmer implements Parcelable, Serializable
 
     public Farmer(JSONObject farmerJsonObject, String extensionPersonnel){
         try{
+            Log.d(TAG, "Json for this farmer = "+farmerJsonObject.toString());
             fullName = farmerJsonObject.getString("name");
             this.extensionPersonnel = extensionPersonnel;
             mobileNumber = farmerJsonObject.getString("mobile_no");
-            longitude = farmerJsonObject.getString("longitude");
-            latitude = farmerJsonObject.getString("latitude");
+            longitude = farmerJsonObject.getString("gps_longitude");
+            latitude = farmerJsonObject.getString("gps_latitude");
             simCardSN = farmerJsonObject.getString("sim_card_sn");
             preferredLocale = farmerJsonObject.getString("pref_locale");
             isInFarm = false;
diff --git a/MistroFarmer/src/main/res/layout/activity_edit_farmer.xml b/MistroFarmer/src/main/res/layout/activity_edit_farmer.xml
index 805517c..ccace5b 100644
--- a/MistroFarmer/src/main/res/layout/activity_edit_farmer.xml
+++ b/MistroFarmer/src/main/res/layout/activity_edit_farmer.xml
@@ -70,31 +70,31 @@
             android:layout_marginLeft="@dimen/activity_f_r_edit_text_margin_left"
             android:textSize="@dimen/normal_text_size"
             android:textColor="@color/text_input_color" />
-        <TextView
-            android:id="@+id/number_of_cows_tv"
-            android:layout_width="wrap_content"
-            android:layout_height="wrap_content"
-            android:layout_marginTop="@dimen/activity_f_r_text_view_margin_top"
-            android:textSize="@dimen/normal_text_size"/>
-        <EditText
-            android:id="@+id/number_of_cows_et"
-            android:layout_width="120dp"
+        <LinearLayout
+            android:layout_width="match_parent"
             android:layout_height="wrap_content"
-            android:layout_marginTop="@dimen/activity_f_r_edit_text_margin_top"
-            android:layout_marginLeft="@dimen/activity_f_r_edit_text_margin_left"
-            android:inputType="number"
-            android:gravity="center_horizontal"
-            android:textColor="@color/text_input_color"
-            android:textSize="@dimen/normal_text_size"/>
-        <Button
-            android:id="@+id/edit_b"
-            android:layout_width="wrap_content"
-            android:layout_height="@dimen/normal_button_layout_height"
-            android:layout_marginTop="15dp"
-            android:layout_gravity="bottom|right"
-            android:paddingLeft="@dimen/normal_button_padding_side"
-            android:paddingRight="@dimen/normal_button_padding_side"
-            android:textSize="@dimen/normal_text_size"/>
+            android:orientation="horizontal">
+            <Button
+                android:id="@+id/cancel_b"
+                android:layout_width="wrap_content"
+                android:layout_height="@dimen/normal_button_layout_height"
+                android:layout_marginTop="15dp"
+                android:layout_gravity="bottom|right"
+                android:paddingLeft="@dimen/normal_button_padding_side"
+                android:paddingRight="@dimen/normal_button_padding_side"
+                android:textSize="@dimen/normal_text_size"
+                android:layout_weight="1"/>
+            <Button
+                android:id="@+id/edit_b"
+                android:layout_width="wrap_content"
+                android:layout_height="@dimen/normal_button_layout_height"
+                android:layout_marginTop="15dp"
+                android:layout_gravity="bottom|right"
+                android:paddingLeft="@dimen/normal_button_padding_side"
+                android:paddingRight="@dimen/normal_button_padding_side"
+                android:textSize="@dimen/normal_text_size"
+                android:layout_weight="1"/>
+        </LinearLayout>
     </LinearLayout>
 </ScrollView>
 
diff --git a/MistroFarmer/src/main/res/layout/activity_farmer_selection.xml b/MistroFarmer/src/main/res/layout/activity_farmer_selection.xml
index 9389a92..4604465 100644
--- a/MistroFarmer/src/main/res/layout/activity_farmer_selection.xml
+++ b/MistroFarmer/src/main/res/layout/activity_farmer_selection.xml
@@ -36,6 +36,13 @@
             android:paddingLeft="@dimen/normal_button_padding_side"
             android:paddingRight="@dimen/normal_button_padding_side"
             android:textSize="@dimen/normal_text_size"/>
+        <Button
+            android:id="@+id/back_b"
+            android:layout_width="@dimen/back_button_width"
+            android:layout_height="50dp"
+            android:layout_marginTop="90dp"
+            android:paddingLeft="15dp"
+            android:paddingRight="15dp"/>
     </LinearLayout>
 
 </ScrollView>
\ No newline at end of file
diff --git a/MistroFarmer/src/main/res/values/strings_en.xml b/MistroFarmer/src/main/res/values/strings_en.xml
index f5e79af..2ac28fc 100644
--- a/MistroFarmer/src/main/res/values/strings_en.xml
+++ b/MistroFarmer/src/main/res/values/strings_en.xml
@@ -319,4 +319,5 @@
     <string name="edit_en">Edit</string>
     <string name="select_farmer_en">Select a farmer</string>
     <string name="select_en">Select</string>
+    <string name="phone_number_en">Phone Number</string>
 </resources>
diff --git a/MistroFarmer/src/main/res/values/strings_kp.xml b/MistroFarmer/src/main/res/values/strings_kp.xml
index 9dd226b..ce32f77 100644
--- a/MistroFarmer/src/main/res/values/strings_kp.xml
+++ b/MistroFarmer/src/main/res/values/strings_kp.xml
@@ -317,4 +317,5 @@
     <string name="edit_kp">Edit</string>
     <string name="select_farmer_kp">Select a farmer</string>
     <string name="select_kp">Select</string>
+    <string name="phone_number_kp">Phone Number</string>
 </resources>
diff --git a/MistroFarmer/src/main/res/values/strings_kr.xml b/MistroFarmer/src/main/res/values/strings_kr.xml
index 5562adb..2618cc0 100644
--- a/MistroFarmer/src/main/res/values/strings_kr.xml
+++ b/MistroFarmer/src/main/res/values/strings_kr.xml
@@ -316,4 +316,5 @@
     <string name="edit_kr">Edit</string>
     <string name="select_farmer_kr">Select a farmer</string>
     <string name="select_kr">Select</string>
+    <string name="phone_number_kr">Phone Number</string>
 </resources>
\ No newline at end of file
diff --git a/MistroFarmer/src/main/res/values/strings_lu.xml b/MistroFarmer/src/main/res/values/strings_lu.xml
index 27d6f60..4f2b899 100644
--- a/MistroFarmer/src/main/res/values/strings_lu.xml
+++ b/MistroFarmer/src/main/res/values/strings_lu.xml
@@ -319,4 +319,5 @@
     <string name="edit_lu">Edit</string>
     <string name="select_farmer_lu">Select a farmer</string>
     <string name="select_lu">Select</string>
+    <string name="phone_number_lu">Phone Number</string>
 </resources>
diff --git a/MistroFarmer/src/main/res/values/strings_nn.xml b/MistroFarmer/src/main/res/values/strings_nn.xml
index c0a8fad..9c679cb 100644
--- a/MistroFarmer/src/main/res/values/strings_nn.xml
+++ b/MistroFarmer/src/main/res/values/strings_nn.xml
@@ -317,4 +317,5 @@
     <string name="edit_nn">Edit</string>
     <string name="select_farmer_nn">Select a farmer</string>
     <string name="select_nn">Select</string>
+    <string name="phone_number_nn">Phone Number</string>
 </resources>
diff --git a/MistroFarmer/src/main/res/values/strings_sw.xml b/MistroFarmer/src/main/res/values/strings_sw.xml
index 853a8dc..31423e5 100644
--- a/MistroFarmer/src/main/res/values/strings_sw.xml
+++ b/MistroFarmer/src/main/res/values/strings_sw.xml
@@ -317,4 +317,5 @@
     <string name="edit_sw">Edit</string>
     <string name="select_farmer_sw">Select a farmer</string>
     <string name="select_sw">Select</string>
+    <string name="phone_number_sw">Phone Number</string>
 </resources>
-- 
2.7.1

